src/sage/rings/species.py

```diff
@@ -0,0 +1,1250 @@
+            """
+            res = super().__pow__(n)
+            elm = self.parent()._cache_get(res)
+            if elm._group is None:
+                elm._elmexp(self, n)
+                elm._canonicalize()
+            return elm
+
+        def _element_key(self):
+            r"""
+            Return the cache lookup key for ``self``.
+            """
+            return self
+
+        @cached_method
+        def _canonicalize(self):
+            r"""
+            Canonicalize this molecular species by sorting the orbits by
+            length and making them consecutive.
+
+            EXAMPLES::
+
+                sage: P = PolynomialSpecies(2)
+                sage: G = PermutationGroup([[(1,2),(3,4),(5,6),(7,8,9,10)]]); G
+                Permutation Group with generators [(1,2)(3,4)(5,6)(7,8,9,10)]
+                sage: H = PermutationGroup([[(1,2,3,4),(5,6),(7,8),(9,10)]]); H
+                Permutation Group with generators [(1,2,3,4)(5,6)(7,8)(9,10)]
+                sage: A = P(G, {1: [1,2,3,4], 2: [5,6,7,8,9,10]})
+                sage: A.support()[0]._dompart
+                ((5, 6, 7, 8), (9, 10, 1, 2, 3, 4))
+                sage: C = P(G, {1: [1,2,5,6], 2: [3,4,7,8,9,10]})
+                sage: C.support()[0]._dompart
+                ((5, 6, 7, 8), (9, 10, 1, 2, 3, 4))
+            """
+            if self._group is None or self._group == SymmetricGroup(0):
+                return
+            sorted_orbits = sorted([sorted(orbit) for orbit in self._group.orbits()], key=len, reverse=True)
+            pi = PermutationGroupElement(list(chain.from_iterable(sorted_orbits))).inverse()
+            self._group = PermutationGroup(gap_group=libgap.ConjugateGroup(self._group, pi))
+            self._dompart = tuple(tuple(pi(k) for k in v) for v in self._dompart)
+
+        def grade(self):
+            r"""
+            Return the grade of ``self``.
+            """
+            return self._tc
+
+        def domain(self):
+            r"""
+            Return the domain of ``self``.
+            """
+            return FiniteEnumeratedSet(range(1, self._tc + 1))
+
+        def cartesian_product(self, other):
+            r"""
+            Compute the cartesian product of ``self`` and ``other``.
+
+            EXAMPLES:
+
+            Exercise 2.1.9 from the BLL book::
+
+                sage: P = PolynomialSpecies(["X"])
+                sage: M = P._indices
+                sage: C3 = M(CyclicPermutationGroup(3))
+                sage: X = M(SymmetricGroup(1))
+                sage: E2 = M(SymmetricGroup(2))
+                sage: C3.cartesian_product(C3) # C3 x C3 = 2*C3
+                2*{((1,2,3),): ((1, 2, 3),)}
+                sage: (X^3).cartesian_product(C3) # X^3 x C3 = 2*X^3
+                2*{((),): ((1,),)}^3
+                sage: (X*E2).cartesian_product(X*E2) # X*E2 x X*E2 = X*E2 + X^3
+                {((),): ((1,),)}*{((1,2),): ((1, 2),)} + {((),): ((1,),)}^3
```

Comment:
using the patch below:
```suggestion
                sage: C3.cartesian_product(C3)
                2*C_3
                sage: (X^3).cartesian_product(C3)
                2*X^3
                sage: (X*E2).cartesian_product(X*E2)
                X*E_2 + X^3
```